WIP: Vicinity generalisations

benowen-bom opened the pull request:
This PR generalises the vicinity CLI to use additional vicinity calculations, namely mean-in-vicinity, min-in-vicinity and std-in-vicinity to enable generation of in-vicinity features for use in the RainForests calibration.

The approach taken is the same as that used for the max-in-vicinity, but generalising the scipy filter function to use the `generic_filter` method with the relevant operation (max, min, mean, std) passed to this function to apply over the neighbourhood patches.

The operator to use is passed in via an optional `operator` kwarg, with the default set to `max`; so the current behaviour of this CLI remains unchanged if called using the existing pattern. There is also an addition `new_name` kwarg which can be set to remain the resultant cube name; this is necessary to distinguish the different vicinity cubes that are used in the RainForests calibration, as the default naming set by `rename_vicinity_cube` has no reference to the operator (and so the result for each operator is indistinguishable).

Testing:

- [ ] Ran tests and they passed OK
- [ ] Added new tests for the new feature(s)

esteban-abellan commented on improver/utilities/spatial.py:
```diff
@@ -716,18 +718,18 @@ def process(self, cube: Cube) -> CubeList[Cube, Cube]:
-    the maximum value within the vicinity is chosen.
-    If a land-mask has been supplied, process land and sea points
+    Evaluate the specified operator over grid points within a defined radius.
+    Where operator applied is evaluates area maxima or minima, all grid points
```
```suggestion
Where the operator is applied, it evaluates area maxima or minima, all grid points
```

esteban-abellan commented on improver/utilities/spatial.py:
```diff
@@ -779,6 +782,152 @@ def maximum_within_vicinity(
+    landmask: Optional[ndarray] = None,
+) -> Union[MaskedArray, ndarray]:
+    """
+    Find mean values over grid points within a defined radius.
```
```suggestion
Find std values over grid points within a defined radius.
```

benowen-bom replied:
Good pickup. I've updated the text, but used standard-deviation rather than std.

benowen-bom commented on improver_tests/utilities/spatial/test_spatial.py:
```diff
@@ -773,7 +776,353 @@ def test_maximum_within_vicinity(grid, radius, landmask, expected_result):
+    ],
+)
+def test_std_within_vicinity(grid, radius, landmask, expected_result):
+    """Test that mean_within_vicinity function returns the expected
```
Fixed.

benowen-bom commented on improver/utilities/spatial.py:
```diff
@@ -757,20 +760,24 @@ def maximum_within_vicinity(
             max_data = np.where(landmask == match, matched_max_data, max_data)
     else:
         # The following command finds the maximum value for each grid point
         # from within a square of length "size"
```
Thanks. I've updated this to make it more generic.

benowen-bom commented on improver/utilities/spatial.py:
```diff
@@ -779,6 +786,152 @@ def maximum_within_vicinity(
+    """
+    Find grid points where a phenomenon occurs within a defined radius.
+    The occurrences within this vicinity are minimised, such that all
+    grid points within the vicinity are recorded as having an occurrence.
```
Good pick-up. My focus has been on realization based data, but this comment is clearly referring to binary values and the comment for the maximum doesn't carry over so well (particularly when what I've said is wrong!). I've updated the comment to be clearer (and accurate in the binary case).

benowen-bom commented on improver/utilities/spatial.py:
```diff
@@ -779,6 +786,152 @@ def maximum_within_vicinity(
+            An array of values to which the process is applied.
+        grid_point_radius:
+            The radius in grid points about each point within which to
+            determine the maximum value.
```
Fixed.

benowen-bom commented on improver/utilities/spatial.py:
```diff
@@ -757,25 +764,199 @@ def maximum_within_vicinity(
-        # from within a square of length "size"
-        max_data = maximum_filter(unmasked_grid, size=grid_points)
+        # The following command finds the value for the specified operation for
+        # each grid point from within a square of length "size"
```
Updated as per suggestion.

bayliffe commented on improver_tests/utilities/spatial/test_spatial.py:
```diff
@@ -773,7 +794,388 @@ def test_maximum_within_vicinity(grid, radius, landmask, expected_result):
+        ),
+        # Vicinity processing, with one non-zero corner value resulting
+        # in neighbouring cells values of 1 within the limit of the
+        # defined vicinity radius
```
I don't quite understand this description though the test seems fine.

benowen-bom replied:
The dangers of copy-paste. Thanks for being so thorough with this!

benowen-bom replied:
Haha! I can see what you mean. I can assure you that the tone here was sincere. The state of this PR is significantly better than before your review, so thank you.

bayliffe reviewed (commented):
Thanks again Ben. I've now gone through the tests as well. I think we're basically there, just a few last bits.

benowen-bom commented:
Thanks @bayliffe for your thorough review; the code is in a much better shape now following your suggestions.

I think I have addressed all the points now, so this is ready for another pass.

bayliffe reviewed (approved):
Looks good Ben. One commented line to remove, but I'll approve this now so you can merge it once that change is made and someone your side reapproves it. Thanks for all the changes.

benowen-bom commented on improver/utilities/spatial.py:
```diff
@@ -1232,7 +1246,8 @@ def process(self, cube: Cube, new_name: str = None) -> Cube:
         # Add cell method to describe the vicinity operation applied.
-        result_cube.add_cell_method(CellMethod(method=self.cell_method, coords="area"))
+        set_vicinity_cell_method(result_cube, operation=self.cell_method)
+        # result_cube.add_cell_method(CellMethod(method=self.cell_method, coords="area"))
```
Done.
